components/suggest/src/weather.rs

```diff
@@ -1,53 +1,235 @@
+#[derive(Debug, Default)]
+pub struct WeatherCache {
+    /// Cached value of the same name from `SuggestProviderConfig::Weather`.
+    min_keyword_length: i32,
+    /// Cached value of the same name from `SuggestProviderConfig::Weather`.
+    score: f64,
+    /// Max length of all weather keywords.
+    max_keyword_length: usize,
+    /// Max word count across all weather keywords.
+    max_keyword_word_count: usize,
+}
+
 impl SuggestDao<'_> {
     /// Fetches weather suggestions.
     pub fn fetch_weather_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
-        // Weather keywords are matched by prefix but the query must be at least
-        // three chars long. Unlike the prefix matching of other suggestion
-        // types, the query doesn't need to contain the first full word.
+        // We'll just stipulate we won't support tiny queries in order to avoid
+        // a bunch of work when the user starts typing a query.
         if query.keyword.len() < 3 {
             return Ok(vec![]);
         }
 
-        let keyword_lowercased = &query.keyword.trim().to_lowercase();
-        let suggestions = self.conn.query_rows_and_then_cached(
+        // The first step in parsing the query is lowercasing and splitting it
+        // into words. We want to avoid that work for strings that are so long
+        // they can't possibly match. The longest possible weather query is two
+        // geonames + one weather keyword + at least two spaces between those
+        // three components, say, 10 spaces total for some wiggle room. There's
+        // not much point in an analogous min length check since weather
+        // suggestions can be matched on city alone and many city names are only
+        // a few characters long ("nyc").
+        let g_cache = self.geoname_cache();
+        let w_cache = self.weather_cache();
+        let max_query_len = 2 * g_cache.max_name_length + w_cache.max_keyword_length + 10;
+        if max_query_len < query.keyword.len() {
+            return Ok(vec![]);
+        }
+
+        let max_win_size =
+            std::cmp::max(g_cache.max_name_word_count, w_cache.max_keyword_word_count);
+        let kw = query.keyword.to_lowercase();
+        let paths = filter_map_chunks::<Token>(&kw, max_win_size, |chunk, chunk_index, path| {
+            // Match the chunk to a token type that hasn't already been matched
+            // in this path.
+            for tt in [
+                TokenType::City,
+                TokenType::Region,
+                TokenType::WeatherKeyword,
+            ] {
+                if !path.iter().any(|t| t.token_type() == tt) {
+                    let tokens = self.match_weather_tokens(tt, path, chunk, chunk_index == 0)?;
+                    if !tokens.is_empty() {
+                        // This chunk matches, so return the tokens to continue
+                        // down this path.
+                        return Ok(Some(tokens));
```

Comment:
I just realized that this short-circuits the logic and doesn't allow for matching other token types.  This means that `rochester ny weather` will work, but `ny rochester weather` won't because `ny` is always matched as a city only and not a region.

Reply:
You're right, really nice catch! It's an error. I'm working on a fix.

Reply:
This is fixed in the latest commit.